src/projection.jl

```diff
@@ -31,114 +68,269 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
     end
 end
 
-# fallback (structs)
-function ProjectTo(x::T) where {T}
-    # Generic fallback for structs, recursively make `ProjectTo`s all their fields
+# Structs
+function generic_projectto(x::T; kw...) where {T}
+    # Generic fallback, recursively make `ProjectTo`s for all their fields
     fields_nt::NamedTuple = backing(x)
-    return ProjectTo{T}(map(ProjectTo, fields_nt))
+    fields_proj = map(fields_nt) do x1
+        if x1 isa Number || x1 isa AbstractArray
+            ProjectTo(x1)
+        else
+            x1
+        end
+    end        
+    # We can't use `T` because if we have `Foo{Matrix{E}}` it should be allowed to make a
+    # `Foo{Diagaonal{E}}` etc. We assume it has a default constructor that has all fields 
+    # but if it doesn't `construct` will give a good error message.
+    wrapT = T.name.wrapper
+    return ProjectTo{wrapT}(; fields_proj..., kw...)
 end
 function (project::ProjectTo{T})(dx::Tangent) where {T}
     sub_projects = backing(project)
     sub_dxs = backing(canonicalize(dx))
-    _call(f, x) = f(x)
-    return construct(T, map(_call, sub_projects, sub_dxs))
-end
-
-# should not work for Tuples and NamedTuples, as not valid tangent types
-function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-    return throw(
-        ArgumentError("The `x` in `ProjectTo(x)` must be a valid differential, not $x")
-    )
+    maybe_call(f::ProjectTo, x) = f(x)
+    maybe_call(f, x) = f
+    return construct(T, map(maybe_call, sub_projects, sub_dxs))
 end
 
 # Generic
+(::ProjectTo{T})(dx::T) where {T} = dx 
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx
+(::ProjectTo{T})(dx::NotImplemented) where {T} = dx
+
+ProjectTo() = ProjectTo{Any}()  # trivial case, exists so that maybe_call(f, x) knows what to do
+(x::ProjectTo{Any})(dx) = dx
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = project(dx.val)  # can't update in-place part
```

Comment:
should we remove this?
Since people are (for now) applying `project` manually, noone would ever want to do this operation manually, since it destroys the `InplaceableThunk` they presumably just spent the time writing.

```suggestion
```

Reply:
It seems there's little downside. I don't like making things too rigid, if (say) you want to try out adding ProjectTo globally, then it's nice if things which can easily work just work, instead of forcing you to make correlated changes to X and Y just to avoid some error case.

Reply:
>  I don't like making things too rigid, 

I do.
This feels like a footgun.

> if (say) you want to try out adding ProjectTo globally, then it's nice if things which can easily work just work, instead of forcing you to make correlated changes to X and Y just to avoid some error case.

That's fine i think.
Someone trying out adding ProjectTo globally needs to be aware of this issue -- that they can't ProjectTo a InplacableThunk, and then they can try to write the code and see that it can't be done in a way that preserves is inplacable nature.

Reply:
If I add it globally to Zygote right now, then I think without this I'd probably need a special branch of ChainRules and any other packages I'm testing with, in which I've gone through and commented out by hand every inplace use. Or maybe not, maybe it depends exactly where in the chain I manage to insert this thing. 

It can be a warning if you think it might silently end up in hot code and break things, no?

Reply:
If you are adding it to Zygote for testing purposes and to debug the idea, you can type-pirate it.
And then deending what you end up with either upstream it.
Or you can introduce a new function that uses `ChainRulesCore.ProjectTo` for most things,
but special cases `InplacableThunk` to do this other thing.

Reply:
It is always nonbreaking to turn an error into a nonerror.
It is often  breaking to change behavour of things that don't error

Reply:
Now it un-thunks and prints a bright red error message.